File: 02_selection_sort/02_selection_sort.js
//SELECTION SORT O(n^2)

// First write a function to find the smallest element in an array
function findSmallestIndex(arr) {
  let smallestElement = arr[0] //Stores the smallest value
  let smallestIndex = 0 //Stores the index of the smallest value
  console.log("before loop " + smallestElement)

  for (let i = 0; i < arr.length; i++) {
    if (arr[i] < smallestElement) {
      smallestElement = arr[i]
      smallestIndex = i
      console.log("during sort loop " + smallestElement)
    }
  } return smallestIndex
}

// Then use the function to write selection sort:
function selectionSort(arr) { //Sorts an array
  let sortedArr = []
  let length = arr.length;
  for (let i = 0; i < length; i++){
    let smallestIndex = findSmallestIndex(arr) // Finds the smallest element in the array
    sortedArr.push(arr.splice(smallestIndex, 1)[0]) // Adds the smallest element to new array
  }
    return sortedArr
}

console.log(selectionSort([5, 3, 6, 2, 10])); // [2, 3, 5, 6, 10]

File: 02_selection_sort/02_selection_sort.rb
# SELECTION SORT O(n^2)

# First write a function to find the smallest element in an array
def find_smallest_index(arr)
  smallest_element = arr[0] # Stores the smallest value
  smallest_index = 0 # Stores the index of the smallest value
  (1...(arr.length)).each do |i|
    if arr[i] < smallest_element
      smallest_element = arr[i]
      smallest_index = i
    end
  end
  smallest_index
end

# Then use the function to write selection sort:
def selection_sort(arr) #Sorts an array
  sorted_arr = []
  arr.length.times do
    smallest_index = find_smallest_index(arr) #Finds the smallest element in the array
    sorted_arr.push(arr.delete_at(smallest_index)) # Adds the smallest element to new array
  end
    sorted_arr
end

p selection_sort([5, 3, 6, 2, 10]); # [2, 3, 5, 6, 10]